service/src/main/java/crawlercommons/urlfrontier/service/AbstractFrontierService.java

```diff
@@ -803,13 +829,18 @@ public void onCompleted() {
     @Override
     public void close() throws IOException {
         closing = true;
-        executorService.shutdown();
+        writeExecutorService.shutdown();
+        readExecutorService.shutdown();
         try {
-            if (!executorService.awaitTermination(500, TimeUnit.MILLISECONDS)) {
-                executorService.shutdownNow();
+            if (!writeExecutorService.awaitTermination(500, TimeUnit.MILLISECONDS)) {
+                writeExecutorService.shutdownNow();
```

Comment:
Maybe we can guard this by another `awaitTermination` call after the request to shutdown was sent, so the thread pool has the chance to terminate properly?

Reply:
hmm, this section already does quite a bit of waiting and being friendly but if it can be made better, we can add a PR to it later

Reply:
Sure